### PR TITLE
Adding Expansion Development Method and more...

### PR DESCRIPTION
**Pull request by Llewellynvdm**
Pull Request for many Issue

### Summary of Changes
- Added a switch to control the message display of the expand feature when cronjob runs
- Added the feature that allows you to setup the behaviour of a selected set of components that will be auto build and installed. So you do not need to manually compile and install those components any more, the system does all that automatically for you. We have called the feature Expansion Development Method, and it can be set in the global settings of JCB under the Development Method tab.
- Fixed the auto backup system, to allow JCB to be backed up automatically
- Tweaked the field database area to insure that users are forced to set the database values
- Fixed a strange mismatch issue with JCB package import with more then 10 components at a time. Also fixed the translation entranslation to source.
- Improved JCB to now be able to have other languages then en-GB as its base/main/global language. This will allow you to build your components in a language of your choice.
- Added edit links to templates and layouts.
- Fixed the alias and title selection issue in admin fields area.
- Added all available languages in Joomla to language area.
- Added minify of JS to the compiler area, also improved the infusion area to make more use of the getCustomScriptBuilder method
- Fixed the Add JavaScript (views-footer) option, since it did not add the javascript, but now does. Also added some spinner to the Joomla Components view.
- Added the option to clone a Component and all its linked data, also to import a JCB package without merging with local Component
- Keeping naming consistency resolved gh-265. Started adding the needed features for the new translation improvement gh-261
- Added the php field properties to a textarea outside of the subform for easier editing

### Testing Instructions
- Check that the demo component is able to compile and that you can all over JCB edit, create and all the other normal functionality still works as expected.
- Check that you can import the demo component without merging, and so have duplicates of all fields, views and other linked data to the component.
-  Test drive JCB with other language as en-GB being the main language.
- Check that the auto back-up system still works.
- Test the Expansion Development Method, it can be set in the global settings of JCB under the Development Method tab.

### Expected result
- That JCB is stable and now having more features and improvements then before.

### CODE REVIEW

File: admin/helpers/componentbuilder.php [hunk 1]
```diff
@@ -2256,6 +2290,607 @@ public static function composerAutoload()
...
+	/**
+	*	To Lock string
+	*
+	*	@param string  $string       The string/array to lock
+	*	@param string  $key          The custom key to use
+	*	@param int      $salt           The switch to add salt and type of salt
+	*	@param int      $dynamic    The dynamic replacement array of salt build string
+	*	@param int      $urlencode  The switch to control url encoding
+	**/
+	public static function lock($string, $key = null, $salt = 2, $dynamic = null, $urlencode = true)
```
Comment (ro-ot) [inline admin/helpers/componentbuilder.php]: This is cool, thanks for sharing!

File: admin/helpers/componentbuilder.php [hunk 2]
```diff
@@ -2256,6 +2290,607 @@ public static function composerAutoload()
...
+	*
+	* 	@param  string      $url               The url of where the task is to be performed
+	*  	@param  array       $_options      The array of curl options/headers to set
+	*	@param  function   $callback      The option to do a call back when task is completed
+	*	@param  int           $threadSize   The size of the thread
+	*
+	* 	@return  bool true   On success
+	*
+	*/
+	public static function curlMultiExec(&$url, &$_options, $callback = null, $threadSize = 20)
```
Comment (ro-ot) [inline admin/helpers/componentbuilder.php]: I like :+1: almost multithreading in PHP wow!

**Comment by ro-ot** [comment]
@Llewellynvdm so far it looks good, I will give you final feedback over the weekend.

File: admin/helpers/componentbuilder.php (hunk 1 above)
Comment (Llewellynvdm) [reply admin/helpers/componentbuilder.php]: Pleasure! simple little trick to lock values for a short period of time.

**Comment by ro-ot** [comment]
I think best let this release go, since there are so many nice fixes in it, and make that feature the next release. But it is really up to you.

**Pull request merged**